File: Assets/Scripts/DouyinSequenceFxBridge.lua
---@var SceneEnvironment       :DouyinSceneEnvironment
---@var OrbitPulseParticle     :UnityEngine.ParticleSystem
---@var OrbitPulseAudio        :UnityEngine.AudioSource
---@var OrbitPulseAudioObject  :UnityEngine.GameObject
---@var EnableOrbitPulseAudio  :bool = true
---@var OrbitPulseAudioVolume  :float = 1
---@var EnableEnvironmentSwitch:bool = false
---@var EnableDebugLog         :bool = true
---@var DayMainLightStrength   :float = 1
---@var NightMainLightStrength :float = 0
---@var DayEmissionIntensity   :float = 0
---@var NightEmissionIntensity :float = 1
---@var DaySkyColor            :Color
---@var DayEquatorColor        :Color
---@var DayGroundColor         :Color
---@var NightSkyColor          :Color
---@var NightEquatorColor      :Color
---@var NightGroundColor       :Color
---@var AmbientMaterials       :UnityEngine.Material[]
---@end

local orbitPulseActive = false
local orbitPulseCycleCount = 0
local orbitPulseTargetReverseBlend = nil

local function Log(message)
    if EnableDebugLog then
        print("[DouyinSequenceFxBridge] " .. tostring(message))
    end
end

local function GetActiveSkyMaterial()
    if SceneEnvironment == nil then
        return nil
    end

    local ok, material = pcall(function()
        return SceneEnvironment.settings.skyMaterial
    end)
    if not ok then
        Log("SceneEnvironment.settings.skyMaterial access failed: " .. tostring(material))
        return nil
    end
    return material
end

local function ApplySkyPulse(reverseBlend, pulseProgress)
    local material = GetActiveSkyMaterial()
    if material == nil then
        return
    end

    material:SetFloat("_ReverseBlend", reverseBlend)
    material:SetFloat("_PulseProgress", pulseProgress)
end

local function GetSceneEnvironmentSettings()
    if SceneEnvironment == nil then
        return nil
    end

    local ok, settings = pcall(function()
        return SceneEnvironment.settings
    end)
    if not ok then
        Log("SceneEnvironment.settings access failed: " .. tostring(settings))
        return nil
    end
    return settings
end

local function Clamp01(value)
    if value == nil then
        return 0
    end
    if value < 0 then
        return 0
    end
    if value > 1 then
        return 1
    end
    return value
end

local function ResolveAudioSource(audioSource, audioObject)
    if audioSource ~= nil then
        return audioSource
    end

    if audioObject ~= nil then
        return audioObject:GetComponent(typeof(CS.UnityEngine.AudioSource))
    end

    return nil
end

local function PlayOrbitPulseAudio()
    if EnableOrbitPulseAudio == false then
        return
    end

    local audioSource = ResolveAudioSource(OrbitPulseAudio, OrbitPulseAudioObject)
    if audioSource == nil then
        return
    end

    local ok, errorMessage = pcall(function()
        audioSource.enabled = true
        audioSource.playOnAwake = false
        audioSource.loop = false
        audioSource.volume = Clamp01(OrbitPulseAudioVolume or 1)
        audioSource:Stop()
        audioSource:Play()
    end)
    if not ok then
        Log("Orbit pulse audio play failed: " .. tostring(errorMessage))
    end
end

local function StopOrbitPulseAudio()
    local audioSource = ResolveAudioSource(OrbitPulseAudio, OrbitPulseAudioObject)
    if audioSource == nil then
        return
    end

    local ok, errorMessage = pcall(function()
        audioSource:Stop()
    end)
    if not ok then
        Log("Orbit pulse audio stop failed: " .. tostring(errorMessage))
    end
end

local function LerpFloat(fromValue, toValue, t)
    return fromValue + (toValue - fromValue) * t
end

local function ResolveColor(value, fallback)
    if value ~= nil then
        return value
    end
    return fallback
end

local function ToUnityColor(value)
    if value == nil then
        return CS.UnityEngine.Color.black
    end

    return CS.UnityEngine.Color(value.r, value.g, value.b, value.a)
end

local function LerpColor(fromValue, toValue, t)
    return CS.UnityEngine.Color.Lerp(ToUnityColor(fromValue), ToUnityColor(toValue), t)
end

local function ApplyAmbientMaterialColors(skyColor, equatorColor, groundColor)
    if AmbientMaterials == nil then
        return
    end

    for i = 0, AmbientMaterials.Length - 1 do
        local material = AmbientMaterials[i]
        if material ~= nil then
            material:SetColor("_SkyColor", skyColor)
            material:SetColor("_EquatorColor", equatorColor)
            material:SetColor("_GroundColor", groundColor)
        end
    end
end

local function ApplyAmbientColors(nightBlend01)
    local skyColor = LerpColor(
        ResolveColor(DaySkyColor, CS.UnityEngine.Color.white),
        ResolveColor(NightSkyColor, CS.UnityEngine.Color.black),
        nightBlend01)
    local equatorColor = LerpColor(
        ResolveColor(DayEquatorColor, CS.UnityEngine.Color.white),
        ResolveColor(NightEquatorColor, CS.UnityEngine.Color.black),
        nightBlend01)
    local groundColor = LerpColor(
        ResolveColor(DayGroundColor, CS.UnityEngine.Color.white),
        ResolveColor(NightGroundColor, CS.UnityEngine.Color.black),
        nightBlend01)

    CS.UnityEngine.RenderSettings.ambientSkyColor = skyColor
    CS.UnityEngine.RenderSettings.ambientEquatorColor = equatorColor
    CS.UnityEngine.RenderSettings.ambientGroundColor = groundColor
    CS.UnityEngine.Shader.SetGlobalColor("_SkyColor", skyColor)
    CS.UnityEngine.Shader.SetGlobalColor("_EquatorColor", equatorColor)
    CS.UnityEngine.Shader.SetGlobalColor("_GroundColor", groundColor)
    ApplyAmbientMaterialColors(skyColor, equatorColor, groundColor)
end

local function ApplyEnvironmentBlend(nightBlend01)
    local settings = GetSceneEnvironmentSettings()
    if settings ~= nil then
        local ok, err = pcall(function()
            settings.mainLightStrength = LerpFloat(DayMainLightStrength or 1, NightMainLightStrength or 0, nightBlend01)
        end)
        if not ok then
            Log("settings.mainLightStrength write failed: " .. tostring(err))
        end
    end

    CS.UnityEngine.Shader.SetGlobalFloat(
        "_EmissionIntensity",
        LerpFloat(DayEmissionIntensity or 0, NightEmissionIntensity or 1, nightBlend01))
    ApplyAmbientColors(nightBlend01)
end

local function ResolveNightBlend01(reverseBlend, pulseProgress)
    local progress = Clamp01(pulseProgress)
    if reverseBlend ~= nil and reverseBlend > 0.5 then
        return 1 - progress
    end
    return progress
end

function ResetToBase()
    orbitPulseActive = false
    orbitPulseTargetReverseBlend = nil
    StopOrbitPulseAudio()

    local material = GetActiveSkyMaterial()
    if material ~= nil then
        ApplySkyPulse(material:GetFloat("_ReverseBlend"), 0)
        ApplyEnvironmentBlend(ResolveNightBlend01(material:GetFloat("_ReverseBlend"), 0))
    end
end

function BeginOrbitFx(totalDuration)
    orbitPulseActive = totalDuration ~= nil and totalDuration > 0

    local material = GetActiveSkyMaterial()
    if material ~= nil then
        local currentReverseBlend = material:GetFloat("_ReverseBlend")
        if orbitPulseCycleCount > 0 then
            orbitPulseTargetReverseBlend = currentReverseBlend > 0.5 and 0 or 1
        else
            orbitPulseTargetReverseBlend = currentReverseBlend
        end
        ApplySkyPulse(orbitPulseTargetReverseBlend, 0)
        ApplyEnvironmentBlend(ResolveNightBlend01(orbitPulseTargetReverseBlend, 0))
    end

    if OrbitPulseParticle ~= nil then
        if OrbitPulseParticle.isPlaying then
            OrbitPulseParticle:Stop()
            OrbitPulseParticle:Clear()
        end
        OrbitPulseParticle:Play()
    end

    PlayOrbitPulseAudio()

    if EnableEnvironmentSwitch then
        Log("EnableEnvironmentSwitch is true, but custom environment switching remains intentionally isolated.")
    end
end

function TickOrbitFx(deltaTime, progress01)
    if not orbitPulseActive then
        return
    end

    local material = GetActiveSkyMaterial()
    if material == nil then
        return
    end

    local reverseBlend = orbitPulseTargetReverseBlend
    if reverseBlend == nil then
        reverseBlend = material:GetFloat("_ReverseBlend")
        orbitPulseTargetReverseBlend = reverseBlend
    end

    local pulseProgress = progress01 or 0
    ApplySkyPulse(reverseBlend, pulseProgress)
    ApplyEnvironmentBlend(ResolveNightBlend01(reverseBlend, pulseProgress))
end

function CompleteOrbitFx()
    if not orbitPulseActive then
        return
    end

    orbitPulseActive = false
    orbitPulseCycleCount = orbitPulseCycleCount + 1
end

function GetOrbitNightBlend01(progress01)
    local reverseBlend = orbitPulseTargetReverseBlend
    if reverseBlend == nil then
        local material = GetActiveSkyMaterial()
        if material ~= nil then
            reverseBlend = material:GetFloat("_ReverseBlend")
        else
            reverseBlend = 0
        end
    end

    return ResolveNightBlend01(reverseBlend, progress01)
end

function GetCurrentNightBlend01()
    local material = GetActiveSkyMaterial()
    if material == nil then
        return 0
    end

    return ResolveNightBlend01(material:GetFloat("_ReverseBlend"), material:GetFloat("_PulseProgress"))
end

File: Assets/Scripts/DouyinSequenceFxBridge.cs
using UnityEngine;

public sealed class DouyinSequenceFxBridge : MonoBehaviour
{
    public Material SkyMaterial;
    public Light DirectionalLight;
    public ParticleSystem OrbitPulseParticle;
    public AudioSource OrbitPulseAudio;
    public GameObject OrbitPulseAudioObject;
    public bool EnableOrbitPulseAudio = true;
    public float OrbitPulseAudioVolume = 1f;
    public bool EnableEnvironmentSwitch;
    public bool EnableDebugLog = true;
    public float DayMainLightStrength = 1f;
    public float NightMainLightStrength;
    public float DayEmissionIntensity;
    public float NightEmissionIntensity = 1f;
    public Color DaySkyColor = Color.white;
    public Color DayEquatorColor = Color.white;
    public Color DayGroundColor = Color.white;
    public Color NightSkyColor = Color.black;
    public Color NightEquatorColor = Color.black;
    public Color NightGroundColor = Color.black;
    public Material[] AmbientMaterials;

    private bool orbitPulseActive;
    private int orbitPulseCycleCount;
    private float? orbitPulseTargetReverseBlend;

    public void ResetToBase()
    {
        orbitPulseActive = false;
        orbitPulseTargetReverseBlend = null;
        StopOrbitPulseAudio();

        Material material = GetActiveSkyMaterial();
        if (material == null)
        {
            return;
        }

        float reverseBlend = GetMaterialFloat(material, "_ReverseBlend", 0f);
        ApplySkyPulse(reverseBlend, 0f);
        ApplyEnvironmentBlend(ResolveNightBlend01(reverseBlend, 0f));
    }

    public void BeginOrbitFx(float totalDuration)
    {
        orbitPulseActive = totalDuration > 0f;

        Material material = GetActiveSkyMaterial();
        if (material != null)
        {
            float currentReverseBlend = GetMaterialFloat(material, "_ReverseBlend", 0f);
            orbitPulseTargetReverseBlend = orbitPulseCycleCount > 0 ? (currentReverseBlend > 0.5f ? 0f : 1f) : currentReverseBlend;
            ApplySkyPulse(orbitPulseTargetReverseBlend.Value, 0f);
            ApplyEnvironmentBlend(ResolveNightBlend01(orbitPulseTargetReverseBlend.Value, 0f));
        }

        if (OrbitPulseParticle != null)
        {
            if (OrbitPulseParticle.isPlaying)
            {
                OrbitPulseParticle.Stop();
                OrbitPulseParticle.Clear();
            }
            OrbitPulseParticle.Play();
        }

        PlayOrbitPulseAudio();

        if (EnableEnvironmentSwitch)
        {
            Log("EnableEnvironmentSwitch is true, but custom environment switching remains intentionally isolated.");
        }
    }

    public void TickOrbitFx(float deltaTime, float progress01)
    {
        if (!orbitPulseActive)
        {
            return;
        }

        Material material = GetActiveSkyMaterial();
        if (material == null)
        {
            return;
        }

        if (!orbitPulseTargetReverseBlend.HasValue)
        {
            orbitPulseTargetReverseBlend = GetMaterialFloat(material, "_ReverseBlend", 0f);
        }

        float reverseBlend = orbitPulseTargetReverseBlend.Value;
        float pulseProgress = Mathf.Clamp01(progress01);
        ApplySkyPulse(reverseBlend, pulseProgress);
        ApplyEnvironmentBlend(ResolveNightBlend01(reverseBlend, pulseProgress));
    }

    public void CompleteOrbitFx()
    {
        if (!orbitPulseActive)
        {
            return;
        }

        orbitPulseActive = false;
        orbitPulseCycleCount++;
    }

    public float GetOrbitNightBlend01(float progress01)
    {
        float reverseBlend;
        if (orbitPulseTargetReverseBlend.HasValue)
        {
            reverseBlend = orbitPulseTargetReverseBlend.Value;
        }
        else
        {
            Material material = GetActiveSkyMaterial();
            reverseBlend = material != null ? GetMaterialFloat(material, "_ReverseBlend", 0f) : 0f;
        }

        return ResolveNightBlend01(reverseBlend, progress01);
    }

    public float GetCurrentNightBlend01()
    {
        Material material = GetActiveSkyMaterial();
        if (material == null)
        {
            return 0f;
        }

        return ResolveNightBlend01(
            GetMaterialFloat(material, "_ReverseBlend", 0f),
            GetMaterialFloat(material, "_PulseProgress", 0f));
    }

    private Material GetActiveSkyMaterial()
    {
        return SkyMaterial;
    }

    private void ApplySkyPulse(float reverseBlend, float pulseProgress)
    {
        Material material = GetActiveSkyMaterial();
        if (material == null)
        {
            return;
        }

        if (material.HasProperty("_ReverseBlend"))
        {
            material.SetFloat("_ReverseBlend", reverseBlend);
        }
        if (material.HasProperty("_PulseProgress"))
        {
            material.SetFloat("_PulseProgress", pulseProgress);
        }
    }

    private void ApplyEnvironmentBlend(float nightBlend01)
    {
        if (DirectionalLight != null)
        {
            DirectionalLight.intensity = Mathf.Lerp(DayMainLightStrength, NightMainLightStrength, nightBlend01);
        }

        Shader.SetGlobalFloat("_EmissionIntensity", Mathf.Lerp(DayEmissionIntensity, NightEmissionIntensity, nightBlend01));

        Color skyColor = Color.Lerp(DaySkyColor, NightSkyColor, nightBlend01);
        Color equatorColor = Color.Lerp(DayEquatorColor, NightEquatorColor, nightBlend01);
        Color groundColor = Color.Lerp(DayGroundColor, NightGroundColor, nightBlend01);

        RenderSettings.ambientSkyColor = skyColor;
        RenderSettings.ambientEquatorColor = equatorColor;
        RenderSettings.ambientGroundColor = groundColor;
        Shader.SetGlobalColor("_SkyColor", skyColor);
        Shader.SetGlobalColor("_EquatorColor", equatorColor);
        Shader.SetGlobalColor("_GroundColor", groundColor);

        if (AmbientMaterials == null)
        {
            return;
        }

        for (int i = 0; i < AmbientMaterials.Length; i++)
        {
            Material material = AmbientMaterials[i];
            if (material == null)
            {
                continue;
            }

            if (material.HasProperty("_SkyColor")) material.SetColor("_SkyColor", skyColor);
            if (material.HasProperty("_EquatorColor")) material.SetColor("_EquatorColor", equatorColor);
            if (material.HasProperty("_GroundColor")) material.SetColor("_GroundColor", groundColor);
        }
    }

    private AudioSource ResolveOrbitPulseAudio()
    {
        return OrbitPulseAudio != null ? OrbitPulseAudio : OrbitPulseAudioObject != null ? OrbitPulseAudioObject.GetComponent<AudioSource>() : null;
    }

    private void PlayOrbitPulseAudio()
    {
        if (!EnableOrbitPulseAudio)
        {
            return;
        }

        AudioSource audioSource = ResolveOrbitPulseAudio();
        if (audioSource == null)
        {
            return;
        }

        audioSource.enabled = true;
        audioSource.playOnAwake = false;
        audioSource.loop = false;
        audioSource.volume = Mathf.Clamp01(OrbitPulseAudioVolume);
        audioSource.Stop();
        audioSource.Play();
    }

    private void StopOrbitPulseAudio()
    {
        AudioSource audioSource = ResolveOrbitPulseAudio();
        if (audioSource != null)
        {
            audioSource.Stop();
        }
    }

    private static float ResolveNightBlend01(float reverseBlend, float pulseProgress)
    {
        float progress = Mathf.Clamp01(pulseProgress);
        return reverseBlend > 0.5f ? 1f - progress : progress;
    }

    private static float GetMaterialFloat(Material material, string propertyName, float fallback)
    {
        return material != null && material.HasProperty(propertyName) ? material.GetFloat(propertyName) : fallback;
    }

    private void Log(string message)
    {
        if (EnableDebugLog)
        {
            Debug.Log("[DouyinSequenceFxBridge] " + message, this);
        }
    }
}


